crates/via-serve-static/src/lib.rs

```diff
@@ -0,0 +1,141 @@
+mod etag;
+mod respond;
+mod static_file;
+mod stream_file;
+
+use bitflags::bitflags;
+use std::{path::Path, sync::Arc};
+use via::{Endpoint, Error, Result};
+
+use crate::respond::{respond_to_get_request, respond_to_head_request};
+
+pub struct ServeStatic<'a, State> {
```

Comment:
I think I'm going to work on providing some mechanism to have either an owned or borrowed version of middleware. This would allow us to essentially write classes in cases where it make sense. In other words, allow users of the public api to write `async fn call(self: Arc<Self>, request: Request, next: Next) -> Result<Response, Error>`. This would give users of the framework to have immutable state local to middleware which would be incredibly useful for keeping implementation details of community packages from leaking into public apis. Still, I want to support middleware that is called by ref to allow for potentially "happy path"  optimizations. I'm thinking along the lines of a `trait IntoMiddleware` and an enum similar to `std::borrow::Cow` that can hold an owned or by ref Middleware. I'm pretty confident that we're able to pull this off without adding too much complexity the public API (i.e a backwards compatible change). If my thinking is correct, we'll be able to re-write the `ServeStatic` struct to clone `Arc<Self>` and contain the `config` struct along with a semaphore that can either be configured or adapt to `ulimit` to fix the bug when files are streamed to client.